hail/src/main/scala/is/hail/backend/service/ServiceBackend.scala

```diff
@@ -277,37 +286,56 @@ class ServiceBackend(
     uploadFunction.get()
     uploadContexts.get()
 
-    val jobGroup = submitJobGroupAndWait(backendContext, parts, token, root, stageIdentifier)
-
+    val (jobGroup, startJobId) =
+      submitJobGroupAndWait(backendContext, parts, token, root, stageIdentifier)
     log.info(s"parallelizeAndComputeWithIndex: $token: reading results")
     val startTime = System.nanoTime()
-    var r @ (err, results) = runAll[Option, Array[Byte]](executor) {
-      /* A missing file means the job was cancelled because another job failed. Assumes that if any
-       * job was cancelled, then at least one job failed. We want to ignore the missing file
-       * exceptions and return one of the actual failure exceptions. */
-      case (opt, _: FileNotFoundException) => opt
-      case (opt, e) => opt.orElse(Some(e))
-    }(None) {
-      (partIdxs, parts.indices).zipped.map { (partIdx, jobIndex) =>
-        (() => readResult(root, jobIndex), partIdx)
-      }
-    }
-    if (jobGroup.state != Success && err.isEmpty) {
-      assert(jobGroup.state != Running)
-      val error =
-        jobGroup.state match {
-          case Failure =>
-            new HailBatchFailure(
-              s"Job group ${jobGroup.job_group_id} for batch ${batchConfig.batchId} failed with an unknown error"
-            )
-          case Cancelled =>
-            new CancellationException(
-              s"Job group ${jobGroup.job_group_id} for batch ${batchConfig.batchId} was cancelled"
-            )
-        }
 
-      r = (Some(error), results)
-    }
+    val r @ (_, results) =
+      jobGroup.state match {
+        case Success =>
+          runAllKeepFirstError(executor) {
+            (partIdxs, parts.indices).zipped.map { (partIdx, jobIndex) =>
+              (() => readPartitionResult(root, jobIndex), partIdx)
+            }
+          }
+        case Failure =>
+          val failedEntries = batchClient.getJobGroupJobs(
+            jobGroup.batch_id,
+            jobGroup.job_group_id,
+            Some(JobStates.Failed),
+          )
+          assert(
+            failedEntries.nonEmpty,
+            s"Job group ${jobGroup.job_group_id} failed, but no failed jobs found.",
+          )
+          val error = readPartitionError(root, failedEntries.head.head.job_id - startJobId)
+          val results =
+            for {
+              successes <- batchClient.getJobGroupJobs(
+                jobGroup.batch_id,
+                jobGroup.job_group_id,
+                Some(JobStates.Success),
+              )
+              job <- successes
+              partIdx = job.job_id - startJobId
+            } yield (readPartitionResult(root, partIdx), partIdx)
```

Comment:
This code is duplicated. Maybe add a little helper above (eg above line 294) to encapsulate this